File: non-kernel/frontend/scripts/verify-art-l7-reject-001.mjs
#!/usr/bin/env node
/**
 * Independent reproduction for ART-L7-REJECT-001 (illegal transition rejection).
 * No website UI dependency.
 *
 *   node non-kernel/frontend/scripts/verify-art-l7-reject-001.mjs
 *   node non-kernel/frontend/scripts/verify-art-l7-reject-001.mjs path/to/ART-L7-REJECT-001.json
 *
 * Exit 0 = independent rejection matches sealed receipt.
 */

import { createHash } from "crypto";
import { readFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const ALLOWED = {
  INGESTED: new Set(["NORMALIZED"]),
  NORMALIZED: new Set(["VERIFIED"]),
  VERIFIED: new Set(["CORRELATED"]),
  CORRELATED: new Set(["ARCHIVED"]),
  ARCHIVED: new Set(),
  CONTESTED: new Set(),
};

function canonicalize(value) {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
  const keys = Object.keys(value).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${canonicalize(value[k])}`).join(",")}}`;
}

function sha256Hex(input) {
  return createHash("sha256").update(input, "utf8").digest("hex");
}

function reduceEvent(state, event) {
  const next = {
    lifecycle_state: state.lifecycle_state,
    records: { ...state.records },
    commit_finalized: state.commit_finalized,
  };
  if (event.event_type === "RECORD_INSERT") {
    if (next.records[event.record_id]) throw new Error("duplicate record_id");
    next.records[event.record_id] = event.payload;
  } else if (event.event_type === "STATE_TRANSITION") {
    const st = next.lifecycle_state;
    const t = event.to_state;
    if (!(t === "CONTESTED" || ALLOWED[st]?.has(t))) {
      const err = new Error(`illegal transition ${st}->${t}`);
      err.code = "ILLEGAL_TRANSITION";
      err.from = st;
      err.to = t;
      throw err;
    }
    next.lifecycle_state = t;
  } else if (event.event_type === "COMMIT_FINALIZED") {
    next.commit_finalized = true;
  } else {
    throw new Error(`unsupported event_type ${event.event_type}`);
  }
  return next;
}

function run(capsule) {
  let state = structuredClone(capsule.initial_state);
  let rejection = null;
  let index = 0;
  for (const event of capsule.events) {
    try {
      state = reduceEvent(state, event);
      index++;
    } catch (e) {
      rejection = {
        rejected: true,
        rejection_index: index,
        rejection_code: e.code || "REJECT",
        rejection_message: e.message,
        from_state: e.from,
        to_state: e.to,
        state_before_reject: state,
      };
      break;
    }
  }
  if (!rejection) {
    return { pass: false, reason: "expected rejection did not occur" };
  }

  const state_before_hash = sha256Hex(canonicalize(rejection.state_before_reject));
  let head = "GENESIS";
  for (const event of capsule.events) {
    head = sha256Hex(`${head}|${canonicalize(event)}`);
  }
  const receipt_payload = {
    version_bundle: capsule.version_bundle,
    event_count: capsule.events.length,
    rejection_index: rejection.rejection_index,
    rejection_code: rejection.rejection_code,
    rejection_message: rejection.rejection_message,
    from_state: rejection.from_state,
    to_state: rejection.to_state,
    state_before_hash,
    attempted_ledger_head: head,
    state_mutated: false,
  };
  const receipt_hash = sha256Hex(canonicalize(receipt_payload));

  const exp = capsule.expected;
  const checks = {
    rejected: rejection.rejected === exp.rejected,
    rejection_index: rejection.rejection_index === exp.rejection_index,
    rejection_code: rejection.rejection_code === exp.rejection_code,
    rejection_message: rejection.rejection_message === exp.rejection_message,
    from_state: rejection.from_state === exp.from_state,
    to_state: rejection.to_state === exp.to_state,
    state_before_hash: state_before_hash === exp.state_before_hash,
    attempted_ledger_head: head === exp.attempted_ledger_head,
    receipt_hash: receipt_hash === exp.receipt_hash,
    state_mutated: exp.state_mutated === false,
  };

  return {
    artifactId: capsule.artifactId,
    pass: Object.values(checks).every(Boolean),
    checks,
    computed: {
      state_before_hash,
      attempted_ledger_head: head,
      receipt_hash,
      rejection_code: rejection.rejection_code,
      rejection_message: rejection.rejection_message,
    },
    expected: exp,
  };
}

const here = dirname(fileURLToPath(import.meta.url));
const defaultPath = join(here, "../data/evidence/artifacts/ART-L7-REJECT-001.json");
const path = process.argv[2] || defaultPath;

let capsule;
try {
  capsule = JSON.parse(readFileSync(path, "utf8"));
} catch (err) {
  console.error("FAIL: cannot read capsule", path, err.message);
  process.exit(1);
}

if (capsule.artifactId !== "ART-L7-REJECT-001") {
  console.error("FAIL: unexpected artifactId", capsule.artifactId);
  process.exit(1);
}

const result = run(capsule);
console.log(JSON.stringify(result, null, 2));
process.exit(result.pass ? 0 : 1);

File: non-kernel/frontend/scripts/verify_art_l7_reject_001.py
#!/usr/bin/env python3
"""Independent Python reproduction for ART-L7-REJECT-001.

No website UI. No Node dependency.
Exit 0 = rejection path matches sealed receipt.
"""
from __future__ import annotations

import hashlib
import json
import sys
from copy import deepcopy
from pathlib import Path

ALLOWED = {
    "INGESTED": {"NORMALIZED"},
    "NORMALIZED": {"VERIFIED"},
    "VERIFIED": {"CORRELATED"},
    "CORRELATED": {"ARCHIVED"},
    "ARCHIVED": set(),
    "CONTESTED": set(),
}


def canonicalize(value) -> str:
    if value is None or not isinstance(value, (dict, list)):
        return json.dumps(value, separators=(",", ":"), ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(canonicalize(v) for v in value) + "]"
    keys = sorted(value.keys())
    return "{" + ",".join(json.dumps(k, ensure_ascii=False) + ":" + canonicalize(value[k]) for k in keys) + "}"


def sha256_hex(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def reduce_event(state: dict, event: dict) -> dict:
    next_state = {
        "lifecycle_state": state["lifecycle_state"],
        "records": dict(state["records"]),
        "commit_finalized": state["commit_finalized"],
    }
    et = event["event_type"]
    if et == "RECORD_INSERT":
        rid = event["record_id"]
        if rid in next_state["records"]:
            raise ValueError("duplicate record_id")
        next_state["records"][rid] = event["payload"]
    elif et == "STATE_TRANSITION":
        st = next_state["lifecycle_state"]
        t = event["to_state"]
        if not (t == "CONTESTED" or t in ALLOWED.get(st, set())):
            err = ValueError(f"illegal transition {st}->{t}")
            err.code = "ILLEGAL_TRANSITION"  # type: ignore[attr-defined]
            err.from_state = st  # type: ignore[attr-defined]
            err.to_state = t  # type: ignore[attr-defined]
            raise err
        next_state["lifecycle_state"] = t
    elif et == "COMMIT_FINALIZED":
        next_state["commit_finalized"] = True
    else:
        raise ValueError(f"unsupported event_type {et}")
    return next_state


def run(capsule: dict) -> dict:
    state = deepcopy(capsule["initial_state"])
    rejection = None
    index = 0
    for event in capsule["events"]:
        try:
            state = reduce_event(state, event)
            index += 1
        except ValueError as e:
            rejection = {
                "rejected": True,
                "rejection_index": index,
                "rejection_code": getattr(e, "code", "REJECT"),
                "rejection_message": str(e),
                "from_state": getattr(e, "from_state", None),
                "to_state": getattr(e, "to_state", None),
                "state_before_reject": state,
            }
            break
    if rejection is None:
        return {"pass": False, "reason": "expected rejection did not occur"}

    state_before_hash = sha256_hex(canonicalize(rejection["state_before_reject"]))
    head = "GENESIS"
    for event in capsule["events"]:
        head = sha256_hex(f"{head}|{canonicalize(event)}")
    receipt_payload = {
        "version_bundle": capsule["version_bundle"],
        "event_count": len(capsule["events"]),
        "rejection_index": rejection["rejection_index"],
        "rejection_code": rejection["rejection_code"],
        "rejection_message": rejection["rejection_message"],
        "from_state": rejection["from_state"],
        "to_state": rejection["to_state"],
        "state_before_hash": state_before_hash,
        "attempted_ledger_head": head,
        "state_mutated": False,
    }
    receipt_hash = sha256_hex(canonicalize(receipt_payload))
    exp = capsule["expected"]
    checks = {
        "rejected": rejection["rejected"] is True and exp["rejected"] is True,
        "rejection_index": rejection["rejection_index"] == exp["rejection_index"],
        "rejection_code": rejection["rejection_code"] == exp["rejection_code"],
        "rejection_message": rejection["rejection_message"] == exp["rejection_message"],
        "from_state": rejection["from_state"] == exp["from_state"],
        "to_state": rejection["to_state"] == exp["to_state"],
        "state_before_hash": state_before_hash == exp["state_before_hash"],
        "attempted_ledger_head": head == exp["attempted_ledger_head"],
        "receipt_hash": receipt_hash == exp["receipt_hash"],
        "state_mutated": exp["state_mutated"] is False,
    }
    return {
        "implementation": "python3",
        "artifactId": capsule["artifactId"],
        "pass": all(checks.values()),
        "checks": checks,
        "computed": {
            "state_before_hash": state_before_hash,
            "attempted_ledger_head": head,
            "receipt_hash": receipt_hash,
            "rejection_code": rejection["rejection_code"],
            "rejection_message": rejection["rejection_message"],
        },
        "expected": exp,
    }


def main() -> int:
    default = Path(__file__).resolve().parent.parent / "data/evidence/artifacts/ART-L7-REJECT-001.json"
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else default
    capsule = json.loads(path.read_text(encoding="utf-8"))
    if capsule.get("artifactId") != "ART-L7-REJECT-001":
        print(json.dumps({"pass": False, "error": "unexpected artifactId"}))
        return 1
    result = run(capsule)
    print(json.dumps(result, indent=2))
    return 0 if result.get("pass") else 1


if __name__ == "__main__":
    raise SystemExit(main())
